keylime/tpm/ec_crypto_helper.py

```diff
@@ -0,0 +1,103 @@
+# pylint: disable=protected-access
+
+from typing import Optional
+
+from cryptography.hazmat import backends
+from cryptography.hazmat.backends.openssl.backend import Backend as OpenSSLBackend
+from cryptography.hazmat.primitives.asymmetric.ec import EllipticCurvePrivateKey, EllipticCurvePublicKey
+
+
+class EcCryptoHelper:
+    instance: Optional["EcCryptoHelper"] = None
+
+    @staticmethod
+    def get_instance() -> "EcCryptoHelper":
+        """Create and return a singleton EcCryptoHelper"""
+        if not EcCryptoHelper.instance:
+            EcCryptoHelper.instance = EcCryptoHelper()
+        return EcCryptoHelper.instance
+
+    def __init__(self) -> None:
+        self.backend = backends.default_backend()
+        if not isinstance(self.backend, OpenSSLBackend):
+            raise Exception("The python cryptography default backend must be using OpenSSL")
+
+        # sanity check the backend
+        if "_lib" not in dir(self.backend):
+            raise Exception("self.backend does not have _lib")
+
+        if "_ffi" not in dir(self.backend):
+            raise Exception("self.backend does not have _ffi")
+
+        lib_dir = dir(self.backend._lib)
+        for func in [
+            "BN_CTX_get",
+            "BN_free",
+            "EC_KEY_get0_group",
+            "EC_POINT_new",
+            "EC_POINT_free",
+            "EC_POINT_mul",
+        ]:
```

Comment:
Ah ... I see now why need to access _lib.

There is a consideration here. `_lib` name is an indication that it is private, and should not be used because it can change in later releases (or disappear)

Seems to me that you want to access openssl directly, via the default backend. In this case I would suggest to use pyOpenSSL, as maybe it is exporting the symbols in a more stable way

Reply:
To avoid another dependency (pyOpenSSL)  just for point multiplication I think I will do the same as this here and use Binding directly: https://github.com/pyca/cryptography/blob/main/src/cryptography/hazmat/backends/openssl/backend.py#L147
Binding is documented and there's a good chance it will not change: https://cryptography.io/en/3.0/hazmat/bindings/openssl/

Reply:
A cool idea. Yes. Using public interfaces is more correct.